diff --git a/src/compiler/translator/Compiler.cpp b/src/compiler/translator/Compiler.cpp
index dfaf27f..369501f 100644
--- a/src/compiler/translator/Compiler.cpp
+++ b/src/compiler/translator/Compiler.cpp
@@ -242,30 +242,36 @@ struct UniformSortComparator
         {
             return false;
         }
-        // First, sort by precision: lowp and mediump are smaller than highp
-        if (firstType.getPrecision() != secondType.getPrecision())
+        // Next sort by precisions
+        // Group uniforms into high-precision and non-high-precision. A non-highp uniform is
+        // considered "smaller" than a highp uniform.
+        const TPrecision firstPrecision  = firstType.getPrecision();
+        const TPrecision secondPrecision = secondType.getPrecision();
+        const bool firstIsHighP          = (firstPrecision == TPrecision::EbpHigh);
+        const bool secondIsHighP         = (secondPrecision == TPrecision::EbpHigh);
+        if (firstIsHighP != secondIsHighP)
         {
-            return firstType.getPrecision() != TPrecision::EbpHigh;
+            return secondIsHighP;
         }
-        // We don't sort highp uniforms. If both uniforms are highp, consider them as equivalent
-        if (firstType.getPrecision() == TPrecision::EbpHigh &&
-            secondType.getPrecision() == TPrecision::EbpHigh)
+        // If both are highp, they are equivalent. Do not reorder them.
+        if (firstIsHighP)
         {
             return false;
         }
-        // If both uniforms are mediump or lowp, we further sort them based on a list of criteria
+        // If we reach here, both uniforms are non-highp. We further sort them based on a list of
+        // criteria
         ASSERT(firstType.getPrecision() != TPrecision::EbpHigh &&
                secondType.getPrecision() != TPrecision::EbpHigh);
-        // criteria 1: sort by arrayness. Non-array element is smaller.
-        if (firstType.isArray() != secondType.isArray())
-        {
-            return !firstType.isArray();
-        }
-        // criteria 2: sort by whether the uniform is a struct. Non-structs is smaller.
+        // criteria 1: sort by whether the uniform is a struct. Non-structs is smaller.
         if ((firstType.getStruct() == nullptr) != (secondType.getStruct() == nullptr))
         {
             return firstType.getStruct() == nullptr;
         }
+        // criteria 2: sort by arrayness. Non-array element is smaller.
+        if (firstType.isArray() != secondType.isArray())
+        {
+            return !firstType.isArray();
+        }
         // criteria 3, non-matrix is smaller than matrix
         if (firstType.isMatrix() != secondType.isMatrix())
         {
